src/rdkafka_cgrp.c

```diff
@@ -3056,17 +3989,77 @@ static void rd_kafka_cgrp_rebalance (rd_kafka_cgrp_t *rkcg,
 
         rd_atomic32_set(&rkcg->rkcg_assignment_lost, assignment_lost);
 
-        /* Remove assignment (async), if any. If there is already an
-         * unassign in progress we don't need to bother. */
-        if (!RD_KAFKA_CGRP_WAIT_REBALANCE_CB(rkcg) &&
-            !(rkcg->rkcg_flags & RD_KAFKA_CGRP_F_WAIT_UNASSIGN)) {
-                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WAIT_UNASSIGN;
-
+        if (!is_cooperative) {
+                /* EAGER case - revoke partitions (which will be followed by
+                 * rejoin, if required (it is under normal operation). */
+                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WAIT_UNASSIGN_CALL;
                 rd_kafka_rebalance_op(
                         rkcg,
                         RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
                         rkcg->rkcg_assignment, reason);
+                return;
+        }
+
+
+        // MH: Maybe this would be better as an error, in case I've missed a
+        //     case. It is ultimately preferable to assert, but if I have made
+        //     an error, it's unlikley to bad to continue here (in fact it's
+        //     likely correct).
+        /* COOPERATIVE case. Under usual circumstances, all partitions are not
+         * revoked unless terminating, or partitions are lost. */
+        rd_assert(terminating ||
```

Comment:
is this compiled into release builds?

if yes, this could potentially be exposed as an error instead in case i've made a logic error. or just removed. having this check here has plenty of value, but we also don't want customer apps crashing because of it.

Reply:
i will use `rd_dassert` and emit an error. the reason being: if this case is hit, it is likely the consumer will be in a valid operating state (so best if a prod consumer does not go down). however it indicates an error in reasoning on my part, which I'd like to know about.

Reply:
and subtle reminderto use ./dev-conf.sh noted.